### PR TITLE
[Feat][Reduction] Add list[int] dim support for multi-dim reduction ops

### LINKED ISSUE #804: [FEAT][Reduction] Enable status promotion for 17 reduction ops (multi-dim gap)
# [FEAT][Reduction] Enable status promotion for 17 reduction ops (multi-dim gap)

### Symptom / Motivation
Discovered during PR #798 ([Feat][Reduction] Migrate remaining 17 reduction ops to spec-conformant interface).

All 17 reduction ops were migrated to the spec-conformant `__init__(*, dtype, dim, keepdim)` interface, but status was reverted from `implemented` to `spec-only` during review because the manifest declares `dim: int | list[int]` while implementations only support `dim: int` (raising `NotImplementedError` for `list`/`tuple`).

Affected ops: sum, mean, amax, amin, prod, var, std, var_mean, argmax, argmin, all, any, count_nonzero, l1_norm, l2_norm, inf_norm, logsumexp.

### Root Cause Analysis
The underlying kernel (`ReduceKernel`, `ArgreduceKernel`, etc.) only supports reduction along a single dimension. Multi-dim reduction requires either kernel-level support or a host-side loop that reduces one dimension at a time. This was out of scope for the interface migration PR.

### Related Files
- `tileops/ops/reduction/reduce.py` — `_SimpleReduceOp`, `_WelfordReduceOp`
- `tileops/ops/reduction/all_op.py`, `any_op.py`, `count_nonzero.py` — logical reduce ops
- `tileops/ops/reduction/l1_norm.py`, `l2_norm.py`, `inf_norm.py` — vector norm ops
- `tileops/ops/reduction/argmax.py`, `argmin.py` — argreduce ops
- `tileops/ops_manifest.yaml` — 17 entries with `status: spec-only`

## Goal
Promote all 17 reduction ops to `status: implemented` by closing the multi-dim gap — either implement `list[int]` dim support or narrow the manifest spec via a manifest-only PR.

## Plan

1. Evaluate feasibility: sequential single-dim reduction loop vs. kernel-level multi-dim
2. If implementing: add `list[int]` dim support by iterating reductions from highest dim to lowest
3. If narrowing spec: submit manifest-only PR changing `dim` type from `int | list[int]` to `int` for ops where multi-dim is not meaningful (e.g., argmax/argmin)
4. Add spec tests for multi-dim cases (where implemented)
5. Flip status to `implemented` and verify validator passes

## Constraints
- Must not regress changes from PR #798
- Manifest changes require a separate manifest-only PR per trust model
- Validator resolver bug (#TBD) should be fixed first to avoid false CI failures during promotion

## Acceptance Criteria
- [ ] AC-1: All 17 reduction ops pass validator at L1 (signature check)
- [ ] AC-2: Status flipped to `implemented` for all promoted ops
- [ ] AC-3: Spec tests cover the newly supported dim types

### PR DESCRIPTION
**Pull request by lcy-seso**
## Summary

Add `list[int]` dim support to reduction ops (reduce, l1_norm, l2_norm, inf_norm, all, any, count_nonzero) via a shared `_multidim.py` helper that iterates single-dim reductions from highest to lowest dim. Softmax-family ops (softmax, log_softmax, logsumexp) retain single-dim only due to their normalization semantics.

Closes #804

## Test plan

- [x] **AC-1**: All 17 reduction ops pass validator at L1 (signature check)
  Executed `python scripts/validate_manifest.py`: all manifest checks passed (26 pre-existing bench warnings unrelated to reduction ops).
- [ ] **AC-2**: Status flipped to implemented for all promoted ops
  DEFERRED: Manifest promotion requires a separate manifest-only PR per trust model. This PR correctly leaves manifest `status` unchanged.
- [x] **AC-3**: Spec tests cover the newly supported dim types
  `python -m pytest tests/ops/test_reduce_multidim.py`: 60 passed. New file adds 60 test nodes with no growth in existing test files.

## Test results

| Metric | Value |
|--------|-------|
| Total  | 60    |
| Passed | 60    |
| Failed | 0     |

## Notes

- AC-2 is intentionally deferred: `ops_manifest.yaml` status promotion requires human review in a separate PR per the trust model (`docs/trust-model.md`).
- The `_multidim.py` helper reduces along dims from highest to lowest, handling negative dims and keepdim correctly.

### CODE REVIEW

File: tileops/ops/reduction/reduce.py [hunk 1]
```diff
@@ -124,6 +124,25 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            # Reduce along the last dim (single-dim path on reshaped tensor).
+            N = x.shape[-1]
+            M = prod(x.shape[:-1])
+            x = x.reshape(M, N)
+            kernel = self._get_or_create_kernel(M, N)
+            N_padded = _align_up(N, ALIGNMENT)
+            if N_padded != N:
+                pv = self._pad_value()
+                pad = (0, N_padded - N)
+                x = F.pad(x, pad) if pv == 0.0 else F.pad(x, pad, value=pv)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/reduction/reduce.py:141]: Reviewer determined this is a false positive:

this is an optional DRY refactor, not a correctness issue. In reduce.py the single-dim and multi-dim paths share only the padding call; the surrounding logic differs materially (single-dim uses bounds-check+movedim+reshape, multi-dim uses normalize_dim+flatten_for_multidim+restore_multidim_shape). No developer code change required for review triage.

Current HEAD: 74f7a5a

File: tileops/ops/reduction/_softmax_base.py [hunk 2]
```diff
@@ -80,10 +81,35 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            y = kernel(x)
+            if N_padded != N:
+                y = y[:, :N] if y.ndim == 2 else y
+            # For same-shape ops (softmax/log_softmax), multi-dim is not
+            # meaningful.  For reduced-dim ops (logsumexp), reshape output.
+            if y.ndim == 2:
+                # Same-shape: unsupported multi-dim for softmax/log_softmax.
+                raise ValueError(
+                    "Multi-dim is not supported for same-shape softmax ops"
+                )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/ops/reduction/_softmax_base.py]: Addressed in e0d5f50.

Evidence:
- `_SoftmaxBaseOp.forward()` now checks `self._supports_multidim` at line 92-96 **before** any kernel creation or compilation
- `_supports_multidim = False` by default (line 43); only `LogSumExpOp` overrides to `True`
- SoftmaxOp/LogSoftmaxOp additionally validate in their own `forward()` (commit 1311c03) for defense-in-depth
- Multi-dim list/tuple dims are rejected immediately with a clear ValueError, no compilation cost

Current HEAD: e0d5f50

File: tests/ops/test_reduce_multidim.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+    assert y.shape == ref.shape, f"shape mismatch: {y.shape} vs {ref.shape}"
+    assert torch.allclose(y, ref, **tol), f"max err: {(y - ref).abs().max()}"
+
+
+# ---------------------------------------------------------------------------
+# Regression: empty dim list must be rejected before reaching the helper
+# ---------------------------------------------------------------------------
+
+
+@pytest.mark.smoke
```
Comment (Ibuki-wind) [inline tests/ops/test_reduce_multidim.py:403]: The new validation regression only covers `dim=[]`, but the helpers other key contract is normalizing negative dims and rejecting duplicates. This should be locked at the public-op level as well: for example, `dim=[-1, 0]` should succeed, while `dim=[1, 1]` should raise. Right now only the helper contract is tested, which is weaker than validating each op entrypoint.

File: tests/ops/test_reduce_multidim.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+    assert y.shape == ref.shape, f"shape mismatch: {y.shape} vs {ref.shape}"
+    assert torch.allclose(y, ref, **tol), f"max err: {(y - ref).abs().max()}"
+
+
+# ---------------------------------------------------------------------------
+# Logical reduce ops: all, any, count_nonzero
+# ---------------------------------------------------------------------------
+
+
+class MultiDimLogicalFixture(FixtureBase):
```
Comment (Ibuki-wind) [inline tests/ops/test_reduce_multidim.py:230]: The multi-dim logical-reduce coverage only uses `float32`. One of the actual new risk areas in this PR is that `all/any/count_nonzero` now run the flattened multi-dim path through `_reduce_2d()`, including the storage-dtype conversion branch for `bool/int32/int64/complex* -> float32`. Single-dim tests do not fully protect that path. Add at least one `bool` case and one `complex` case here.

File: tests/ops/test_reduce_multidim.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+        return {"atol": 1e-4, "rtol": 1e-4}
+    return {"atol": 1e-2, "rtol": 1e-2}
+
+
+# ---------------------------------------------------------------------------
+# Simple reduce ops: sum, mean, amax, amin
+# ---------------------------------------------------------------------------
+
+
+@MultiDimFixture
```
Comment (Ibuki-wind) [inline tests/ops/test_reduce_multidim.py:71]: ProdOp also gains multi-dim support through `reduce.py`, but this suite only covers `sum/mean/amax/amin`. `prod` is the one simple-reduce branch with a different padding identity (`1.0`), so it needs its own `dim=list[int]` regression. Without that, a break in the multi-dim `prod` path will not be caught here.

**Review by Ibuki-wind — COMMENTED** [review]
Implementation looks functionally sound from the exercised paths; I did not reproduce a behavior bug locally. The remaining issues are coverage gaps: multi-dim `ProdOp` is untested, logical multi-dim dtype-conversion branches are untested, and public-op validation coverage is still incomplete for negative/duplicate dims. Please close those gaps before considering this series stable.

File: tests/ops/test_reduce_multidim.py (hunk 3 above)
Comment (lcy-seso) [reply tests/ops/test_reduce_multidim.py:403]: Adopted. Added `test_negative_dims_accepted` (SumOp with `dim=[-1, 0]`) and `test_duplicate_dims_raises` (SumOp with `dim=[1, 1]`) at the op entrypoint level. See 8416e57.

File: tests/ops/test_reduce_multidim.py (hunk 4 above)
Comment (lcy-seso) [reply tests/ops/test_reduce_multidim.py:230]: Adopted. Added `torch.bool` param to `MultiDimLogicalFixture` and `MultiDimCountFixture`, with input generation via `torch.randint` for bool dtype. This exercises the `_UNSUPPORTED_STORAGE_DTYPES` → `to_logical_float32` conversion branch in the multi-dim path. See 8416e57.

File: tests/ops/test_reduce_multidim.py (hunk 5 above)
Comment (lcy-seso) [reply tests/ops/test_reduce_multidim.py:71]: Adopted. Added `test_prod_multidim` covering multi-dim product reduction. Uses sequential single-dim `torch.prod` as reference since PyTorch doesn't support `list[int]` for prod. See 8416e57.

File: tests/ops/test_reduce_multidim.py [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+    assert y.shape == ref.shape, f"shape mismatch: {y.shape} vs {ref.shape}"
+    assert torch.allclose(y, ref, **tol), f"max err: {(y - ref).abs().max()}"
+
+
+# ---------------------------------------------------------------------------
+# Logical reduce ops: all, any, count_nonzero
+# ---------------------------------------------------------------------------
+
+
+class MultiDimLogicalFixture(FixtureBase):
```
Comment (Ibuki-wind) [inline tests/ops/test_reduce_multidim.py:230]: This only partially closes the earlier coverage gap. Adding `torch.bool` is useful, but the multi-dim logical path still lacks a `complex64/complex128` case. That path also goes through `_UNSUPPORTED_STORAGE_DTYPES -> to_logical_float32`, and complex truthiness semantics are distinct enough that the original risk is still not fully covered. Please add at least one complex multi-dim case for `all/any/count_nonzero` before considering this closed.

Comment (lcy-seso) [reply tests/ops/test_reduce_multidim.py:230]: Adopted. Added `complex64` param to both `MultiDimLogicalFixture` and `MultiDimCountFixture`, covering the `_UNSUPPORTED_STORAGE_DTYPES -> to_logical_float32` conversion for complex truthiness (nonzero if real or imag != 0). Extracted `_make_logical_input` helper for dtype-aware input generation. See 5610ded.

File: tests/ops/test_reduce_multidim.py [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+                pytest.param(
+                    (4, 32, 256), [0, 1], True, torch.float32,
+                    marks=pytest.mark.full,
+                ),
+                # bool: exercises storage-dtype conversion (bool -> float32)
+                pytest.param(
+                    (4, 32, 256), [0, 1], False, torch.bool,
+                    marks=pytest.mark.full,
+                ),
+                # complex64: distinct truthiness (nonzero if real or imag != 0)
```
Comment (Ibuki-wind) [inline tests/ops/test_reduce_multidim.py:248]: Resolved on `5610ded`: `complex64` multi-dim cases were added for `all/any/count_nonzero`, which closes the remaining `_UNSUPPORTED_STORAGE_DTYPES -> to_logical_float32` coverage gap I called out.
